File: airbyte-commons/src/main/java/io/airbyte/commons/auth/WorkspaceAuthRole.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.auth;

import java.util.Comparator;
import java.util.EnumSet;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public enum WorkspaceAuthRole implements AuthRoleInterface {

  WORKSPACE_ADMIN(500, AuthRoleConstants.WORKSPACE_ADMIN),
  WORKSPACE_EDITOR(400, AuthRoleConstants.WORKSPACE_EDITOR),
  WORKSPACE_RUNNER(300, AuthRoleConstants.WORKSPACE_RUNNER),
  WORKSPACE_READER(200, AuthRoleConstants.WORKSPACE_READER),
  NONE(0, AuthRoleConstants.NONE);

  private final int authority;
  private final String label;

  WorkspaceAuthRole(final int authority, final String label) {
    this.authority = authority;
    this.label = label;
  }

  @Override
  public int getAuthority() {
    return authority;
  }

  @Override
  public String getLabel() {
    return label;
  }

  /**
   * Builds the set of roles based on the provided {@link WorkspaceAuthRole} value.
   * <p>
   * The generated set of auth roles contains the provided {@link WorkspaceAuthRole} (if not
   * {@code null}) and any other authentication roles with a lesser {@link #getAuthority()} value.
   * </p>
   *
   * @param authRole An {@link WorkspaceAuthRole} (may be {@code null}).
   * @return The set of {@link WorkspaceAuthRole} labels based on the provided
   *         {@link WorkspaceAuthRole}.
   */
  public static Set<String> buildWorkspaceAuthRolesSet(final WorkspaceAuthRole authRole) {
    final Set<WorkspaceAuthRole> authRoles = EnumSet.noneOf(WorkspaceAuthRole.class);

    if (authRole != null) {
      authRoles.add(authRole);
      authRoles.addAll(Stream.of(values())
          .filter(role -> !NONE.equals(role))
          .filter(role -> role.getAuthority() < authRole.getAuthority())
          .collect(Collectors.toSet()));
    }

    // Sort final set by descending authority order
    return authRoles.stream()
        .sorted(Comparator.comparingInt(WorkspaceAuthRole::getAuthority))
        .map(WorkspaceAuthRole::getLabel)
        .collect(Collectors.toCollection(LinkedHashSet::new));
  }

}

File: airbyte-commons-auth/src/main/kotlin/io/airbyte/commons/auth/roles/WorkspaceAuthRole.kt
/*
 * Copyright (c) 2020-2026 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.auth.roles

enum class WorkspaceAuthRole(
  private val authority: Int,
  private val label: String,
) : AuthRoleInterface {
  WORKSPACE_ADMIN(500, AuthRoleConstants.WORKSPACE_ADMIN),
  WORKSPACE_EDITOR(400, AuthRoleConstants.WORKSPACE_EDITOR),

  // WORKSPACE_SOURCE_EDITOR and WORKSPACE_DESTINATION_EDITOR are each "workspace editor minus the
  // opposite actor type". They deliberately share an authority so that neither implies the other,
  // while WORKSPACE_EDITOR and above imply both.
  WORKSPACE_SOURCE_EDITOR(350, AuthRoleConstants.WORKSPACE_SOURCE_EDITOR),
  WORKSPACE_DESTINATION_EDITOR(350, AuthRoleConstants.WORKSPACE_DESTINATION_EDITOR),
  WORKSPACE_RUNNER(300, AuthRoleConstants.WORKSPACE_RUNNER),
  WORKSPACE_READER(200, AuthRoleConstants.WORKSPACE_READER),
  NONE(0, AuthRoleConstants.NONE),
  ;

  override fun getAuthority(): Int = authority

  override fun getLabel(): String = label

  companion object {
    /**
     * Builds the set of roles based on the provided [WorkspaceAuthRole] value.
     *
     * The generated set of auth roles contains the provided [WorkspaceAuthRole] (if not `null`)
     * and any other authentication roles with a lesser [getAuthority] value.
     *
     * @param authRole A [WorkspaceAuthRole] (may be `null`).
     * @return A set of [WorkspaceAuthRole] labels based on the provided role.
     */
    @JvmStatic
    fun buildWorkspaceAuthRolesSet(authRole: WorkspaceAuthRole?): Set<String> {
      val authRoles = mutableSetOf<WorkspaceAuthRole>()

      if (authRole != null) {
        authRoles.add(authRole)
        authRoles.addAll(
          entries.filter { it != NONE && it.getAuthority() < authRole.getAuthority() },
        )
      }

      return authRoles
        .sortedBy { it.getAuthority() }
        .map { it.getLabel() }
        .toCollection(LinkedHashSet())
    }
  }
}
